tracer/src/Datadog.Trace/Debugger/PInvoke/DebuggerNativeMethods.cs

```diff
@@ -70,6 +70,8 @@ private static partial class Windows
                 int methodProbesLength,
                 [In] NativeLineProbeDefinition[] lineProbes,
                 int lineProbesLength,
+                [In] NativeSpanProbeDefinition[] spanProbes,
```

Comment:
⚠️ Issue: In order to safely add span probes, you will need to create a new set of PInvokes with the new parameters and leave the existing InstrumentProbes definition as-is.

Background: Because users can add the Datadog.Trace and also install the .NET Tracer via platform packages, we can end up with mismatching `Datadog.Trace.dll` versions at runtime. One issue this causes is that PInvokes cannot be updated across `Datadog.Trace` versions, or else we risk failures when calling into native.

Reply:
To illustrate the scenario a little more, let's assume we make this change and release it in v2.28.0.

Scenario:
- User has .NET Core app with assembly reference to this latest v2.28.0 Datadog.Trace
- User has set up automatic instrumentation with an older version (e.g. v2.25.0) of Datadog.Trace 
- If dynamic instrumentation is enabled, then at startup we will load the v2.28.0 Datadog.Trace (due to the direct assembly reference) and make a PInvoke using the modified signature from this PR
- When the PInvoke calls into the v2.25.0 native side, the signatures do not match up, and we fail the call with an exception

So to solve this scenario as well, we need to make sure the native method exposed via PInvoke is never removed

Reply:
I understand this sentiment but do note that the Dynamic Instrumentation product is currently in **Private Beta** stage. The concern of breaking previous P/Invoke is no relevant until we'll be in GA so it's fine for us to break our own P/Invoke endpoints as well as changing our invokers as we desire.

Reply:
Okay that seems perfectly reasonable for Private Beta 👍🏼 I'm used to raising alarms for this in the Tracer since that component is GA, my bad 😅

Reply:
No problem, thanks for taking a look at this anyhow! It's better to raise alarm, even if it turns out to be false positive.
We understand the pain you described loud and clear and we plan to have a thorough discussions around it prior to turning into GA.
For now, we live in this pink world where we don't have to care about the versioning mismatch whatsoever.

Reply:
One thing to be aware of: _you_ may be in the pink world, but the tracer in general _isn't_. 😢 

There's a _lot_ of version conflict out there. That means that when you announce public beta/GA, you'll almost certainly find someone that blows up because of this. Because they'll update _one_ of the libraries to the required version (if you're lucky 😅), but they will still have the version conflict and 💥

Not saying that it means you definitely _shouldn't_ do this. Just that I don't know if I would 😉 And if history is anything to go by, to expect a long tail of escalations :try_not_to_cry:

Reply:
@andrewlock @zacharycmontoya Just to add a bit more context on why we think it's all pink: the code path that leads to these `PInvokes` will only get triggered based on instrumentation requests that the client library receives through Remote Config. During the private beta, and at the moment we GA, we are free tell our customers that they need to upgrade the version they're using (with automatic instrumentation), but we are not just relying on the hope that they will actually upgrade to prevent the version mismatch, we also:

- Change our backend so that the older tracer versions will not receive instrumentation requests, thus ensuring that from that moment on, the code path that leads to these PInvokes in the old tracer versions will be dead
- Change our UI to tell our customers that they need to upgrade their tracer in order to keep using the product (while showing them a list of which hosts/containers they need to upgrade the tracer on).

I *think* this mitigation should be good enough, unless I'm missing something?

Reply:
> Update our backend so that the older tracer versions will not receive instrumentation requests, thus ensuring that from that moment on, the code path that leads to these PInvokes in the old tracer versions will be dead

This seems like the key point to me. In a version mismatch scenario, both libraries will (theoretically) be receiving remote config, so (theoretically) _both_ would call the PInvokes. Except with you're able to prevent the _older_ version from receiving the remote config call.

What if the auto-instrumentation version is the older one (so it has the older native binary) and they have a _newer_ NuGet package. Are you able to detect that mismatch ahead of time, and prevent _either_ RCM requests in that case? 🤔

Reply:
@andrewlock Thanks a lot for digging deeper and helping us validate our assumptions!

> This seems like the key point to me. In a version mismatch scenario, both libraries will (theoretically) be receiving remote config

Admittedly I hadn't tested this - but are you sure that's theoretically possible? My hypothesis was that since Remote Configuration is only started through `Instrumentation.Initialize`, only the version used in automatic instrumentation will query Remote Config, since I assumed that code path [only gets called via automatic instrumentation](https://github.com/DataDog/dd-trace-dotnet/blob/b066e3cea0d8e26416cc4ffc94e08456e78c27d3/tracer/src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs#L66)?

> What if the auto-instrumentation version is the older one (so it has the older native binary) and they have a newer NuGet package. Are you able to detect that mismatch ahead of time, and prevent either RCM requests in that case? 🤔

 My hypothesis here (again, admittedly untested) is that `Instrumentation.Initialize` will always get called on the version of `Datadog.Trace` that is shipped with automatic instrumentation, and therefore, the [tracer version that gets reported](https://github.com/DataDog/dd-trace-dotnet/blob/9fb7812aeb27b71724a5f34e310e3518dccff81a/tracer/src/Datadog.Trace/RemoteConfigurationManagement/RemoteConfigurationManager.cs#L87) in this example would be the older one, and therefore we will not receive any Dynamic Instrumentation configuration records.